pgxn/neon/walproposer.h

```diff
@@ -282,8 +282,33 @@ typedef struct PageserverFeedback
 	uint32		shard_number;
 } PageserverFeedback;
 
+typedef struct
+{
+	/* Cast to XLogRecPtr */
+	pg_atomic_uint64	flushLsn;
+	/* Cast to SafekeeperState */
+	pg_atomic_uint64	state;
+
+	/*
+	 * connection string for connecting/reconnecting.
+	 *
+	 * May contain private information like password and should not be logged.
+         * It's safe to just have the pointers here in shmem because they all point
+         * to GUCs, which will be at the same address in all backends.
+	 */
+	char	conninfo[MAXCONNINFO];
+	char const *host;
+	char const *port;
+} SafekeeperShmemState;
+
 typedef struct WalproposerShmemState
 {
+	int n_safekeepers;
+	/* Cast to XLogRecPtr */
+	pg_atomic_uint64	commitLsn;    
+	pg_atomic_uint64	epochStartLsn;
+
+	SafekeeperShmemState	safekeeper[MAX_SAFEKEEPERS];
```

Comment:
A compile of questions:
- why do we need two use atomics here?
- why do we need to keep safekeeper state in shared memory?

Frankly speaking I do not completely understand your approach.
My vision is the following:
1. Page server (basebackup.rs calculates `restart_lsn` needed for replication slots)
2. Safekeeper's sync finds donor - most advanced safekeeper and streams WAL from it to other safekeepers through WAL proposer.
3. Wal proposer request WAL from donor using LSN from basebackup
4. Once WAL is downloaded it is possible to start wal sender to replicate this WAL to walreceivers (logical replication)

As far as understand the problem was that there was np sync between 3 and 4.
Also instead of copying all WAL to compute and waiting until it is completed, it is possible to implement on-demand WAL download.  And looks like this PR is implementing this approach. 
But why then it contains some logic of choosing donor? Why it needs to maintain state of all SKs?
Does it eliminate SK sync stage (looks like not)?

Reply:
Yes, this is implementing second approach. As far as I understand, donor is the safekeeper with the most advanced flush LSN. So if we notice that we don't have a piece of WAL locally, we can download it from this most advanced safekeeper. 

Safekeeper state is moved into shmem because walsender is in a separate process, so in order to know which safekeeper is available to download from we need to move information about it into shmem.

Reply:
Does not eliminate SK sync stage

Reply:
> Safekeeper state is moved into shmem because walsender is in a separate process, so in order to know which safekeeper is available to download from we need to move information about it into shmem.

Ok, it's clear.
But I still wonder why do we need atomics and array with each safekeeper state.
It seems to me that either each walsender fills its own cell of safekeeper's array and no synchronisation is needed in this case. Either each walsender aggregate's it data (in this case critical section or atomics are needed).
But you use both.

Reply:
It's because walproposer seems to be the one coordinating elections and receiving feedback from safekeepers. I'm not sure how it would work if we had walsenders doing this as well (when reading WAL, we need to request at a given LSN and term). I think atomics are simpler than having walsender read these things from safekeepers.

Reply:
Regarding why atomics AND spinlock, I didn't touch the spinlock. It was there when I got here, but I didn't need it because atomics suffice :)